### ISSUE TITLE
Final linear layer weights of PatchTST

### ISSUE THREAD

**Issue by gewenkai** [dongbeank/CATS#4]
Hello, I would like to ask why the final linear layer of the original PatchTST in Figure 2 appears to be very structured, while the visualization of the final linear layer in my PatchTST experiments looks chaotic?Should the visualization of the final linear layer in the paper be replaced with attention visualization?

**Comment by dongbeank** [comment]
Did you visualize with the absolute value of linear layer? Just with the vanilla value of linear layer doesn't mean how important that period is.
If you visualized with the absolute value and can't see the figure like in the paper, can you explain which dataset you used? The difference might be related to the dataset characteristics, training hyperparameters, or model convergence state.

**Comment by gewenkai** [comment]
Yes, I used the absolute value and tried the ETTh1, ETTh2, ETTm1, and ETTm2 datasets, but I couldn't obtain similar images. Could you tell me what hyperparameters and datasets were used in Figure 2?

**Comment by dongbeank** [comment]
For the Figure 2(a), I used 24 patch size and 24 stride, and for the Figure 2(b) I used same parameter with the original PatchTST paper. I think it doesn't work for ETTh datasets because periodicity of ETTh datasets is 24 which is too similar with the patch size and stride.
For better distinguishing the linear weights, you might need to adjust the colormap range to avoid having the intermediate colors become indistinguishable due to outliers. Since Python's colormap uses the min and max values for the range, the presence of outliers can compress the visualization of the majority of the data into a narrow color band.

**Comment by dongbeank** [comment]
For Figure 2(a), I also adjusted the input length from 512 to 504 to make it a multiple of the patch length 24.

**Comment by gewenkai** [comment]
Are you sure that the patch length in Figure 2(a) is 24? The paper states that the patch length is 16 with a stride of 8. Could you provide the visualization code? I just conducted an experiment on ETTm following your suggestion with an input length of 504, a patch length of 24, and a stride of 24, but I still couldn't achieve the results presented in the paper. However, my attention weight matrix exhibits a pattern similar to Figure 2, regardless of the dataset used.

**Comment by gewenkai** [comment]
![Image](https://github.com/user-attachments/assets/14e73fed-4bb1-4c7e-906b-a9ccb869f238)

**Comment by dongbeank** [comment]
The patch lengths in my previous response were switched - you're right about that.
For Figure 2(a), the paper states a patch length of 16 with stride of 8, while Figure 2(b) uses 24 for both parameters.
How about using a forecasting horizon of 336? Longer forecasting horizons reveal more vivid patterns. With only 96 outputs, the model might not capture the overall pattern in the output sequence effectively.
Here's the codes (model means PatchTST) 
  weight_matrix = model.model.head.linear.weight.detach().to('cpu').numpy()
  plt.xticks([])
  plt.yticks([])
  plt.imshow(np.abs(weight_matrix.T), aspect='auto', cmap='viridis', vmax=0.017, vmin=0.000) # change vmax which shows better patterns)
  plt.xlabel('Output features')
  plt.ylabel('Input features')
  plt.savefig('patch_linear.pdf', bbox_inches='tight')
  plt.show()

**Comment by gewenkai** [comment]
It's really strange. Despite using the code and parameter settings you provided, I still can't reproduce the results. So, I still believe that the result in Figure 2 should not be the final linear layer weights. If it's a single linear layer, the weights of the linear layer would indeed show the result in your Figure 2 (I've experimented with this, and it really works that way), because a linear layer can capture periodicity. However, in your Figure 2, it's PatchTST, which means there is an Attention part and a feed-forward part before the final linear layer. If the final linear layer weights are still like this, then what have the Attention and feed-forward parts learned? Therefore, I believe the final linear layer weights in PatchTST should be different from those in a model using just a single linear layer for prediction.

**Comment by dongbeank** [comment]
I just reran the PatchTST code with the ETTm1 dataset, 504 seq len, 336 pred len, 24 stride and patch len, and other parameters are the same as in the original paper (include optimizing techniques). I attached the linear layer visualization I just got.
Also, please verify if you compared weights from the early stopping point of the model. The original PatchTST code uses 100 epochs with 20 patience, which could potentially cause overfitting. Check this.
The reason this figure was mentioned in the paper was to show that self-attention actually hinders capturing temporal dependency. The model tends to preserve temporal dependency for better prediction performance even after passing through the encoder layer, despite self-attention mixing the data and potentially disrupting this structure. When the model was replaced with linear, it maintained strong temporal dependency and achieved improved performance. This provides evidence that preserving temporal dependency is important and why still there is a temporal dependency in final linear layer in PatchTST.

[patch_linear.pdf](https://github.com/user-attachments/files/19409564/patch_linear.pdf)

**Comment by zhujiayi18161778290** [comment]
Hello, I also have some doubts about this. Could you directly share the checkpoint weight file you mentioned for the ETTm1 dataset with an input length of 504? I would like to experiment with it as well.

**Comment by gewenkai** [comment]
I couldn't reproduce the results with my current weight file. Perhaps you could reach out to the author and ask if they can provide the checkpoint file.

**Comment by dongbeank** [comment]
Sorry for your confusion and for responding without checking the important details.
The visualization in Figure 2 used PatchTST with **1 encoder layer** to maintain comparable computational complexity with the linear embedding model.
Looking at Table 1, these results were also produced with layer=1, which actually performs better than the original paper's implementation with layer=3. This is because one layer of attention disrupts temporal patterns less than three layers would. That's why your experiments show a more chaotic visualization (again, sorry about that).
You will be able to reproduce the visualization in the paper by using 1 encoder layer in your PatchTST implementation.

**Comment by dongbeank** [comment]
Sorry for your confusion and for responding without checking the important details.
The visualization in Figure 2 used PatchTST with 1 encoder layer to maintain comparable computational complexity with the linear embedding model.
Looking at Table 1, these results were also produced with layer=1, which actually performs better than the original paper's implementation with layer=3. This is because one layer of attention disrupts temporal patterns less than three layers would. That's why your experiments show a more chaotic visualization (again, sorry about that).
I've uploaded model checkpoints for both the overlap and non-overlap versions for your reference.
Additionally, I tried various experiments and found that these temporal patterns could sometimes be maintained during training due to small variations in the training process, but often they weren't. Rather than claiming this happens in all cases, it might be better to use this experiment as an intuitive example demonstrating that maintaining temporal patterns (as in Fig 2c) leads to better performance.

[checkpoints.zip](https://github.com/user-attachments/files/19423735/checkpoints.zip)
